charms/worker/k8s/src/token_distributor.py

```diff
@@ -36,18 +36,86 @@ class ClusterTokenType(Enum):
     NONE = ""
 
 
+class TokenCollector:
+    """Helper class for collecting tokens for units in a relation."""
+
+    def __init__(self, charm: ops.CharmBase, node_name: str):
+        """Initialize a TokenCollector instance.
+
+        Args:
+            charm (CharmBase): A charm object representing the current charm.
+            node_name (str): current node's name
+        """
+        self.charm = charm
+        self.node_name = node_name
+
+    def joined(self, relation: ops.Relation) -> bool:
+        """Report if this unit has completed token transfer.
+
+        Args:
+            relation (ops.Relation): The relation to check
+
+        Returns:
+            bool: if the local unit is joined on this relation.
+        """
+        data = relation.data[self.charm.unit].get("joined")
+        return data == self.node_name
+
+    def request(self, relation: ops.Relation):
+        """Ensure this unit is requesting a token.
+
+        Args:
+            relation (ops.Relation): The relation on which to request
+        """
+        # the presence of node-name is used to request a token
+        relation.data[self.charm.unit]["node-name"] = self.node_name
+
+    @contextlib.contextmanager
+    def recover_token(self, relation: ops.Relation):
+        """Request, recover token, and acknowledge token once used.
+
+        Args:
+            relation (ops.Relation): The relation to check
+
+        Yields:
+            str: extracted token content
+        """
+        self.request(relation)
```

Comment:
I REALLY like this pattern for requesting/recovering the token. It makes the code more readable.